### PR TITLE
Add backend functions allowing WFS analysis

### PR DESCRIPTION
**Pull request by JorgeMartinezG**
This PR leverages current `/stats` endpoint to include analysis between a WFS layer with vector polygons and geotiff raster.

**Features:**
- The api endpoint checks if there is a `wfs_url` dictionary. If yes, it validates the required parameters and makes a WFS request.
- Zonal calculation function checks for a wfs geojson. If yes, statistics are computed based on the intersection between the administrative boundaries and the geojson geometries.
- Within the response, the `geometry` field is included for each zone, which is needed to be rendered in the response layer.

**How to test:**
Using a request client (Paw, Postman, etc) execute the following request to your testing instance. Tested in myanmar tropical storm vector and population raster layers
```
curl -X POST \
  http://localhost:5000/stats \
  -H 'Content-Type: application/json' \
  -d '{
	"geotiff_url": "https://odc.ovio.org/?service=WCS&request=GetCoverage&version=1.0.0&coverage=wp_pop_cicunadj&crs=EPSG%3A4326&bbox=92.2%2C9.7%2C101.2%2C28.5&width=1098&height=2304&format=GeoTIFF",
	"zones_url": "https://prism-admin-boundaries.s3.us-east-2.amazonaws.com/mmr_admin_boundaries.json",
	"group_by": "TS",
	"wfs_params": {
		"url": "https://geonode.wfp.org/geoserver/ows",
		"layer_name": "wld_gdacs_tc_events_buffers",
		"time": "2016-05-18"
	}
}'
```

### CODE REVIEW

**Comment by dubwalla** [comment]
I tested this using the curl statement above. The response I get has null values for all stats. A better example to test is with "time": "2017-04-15". This returned multiple townships as MultiPolygons, each with stats as expected. I'm not fully sure how to test the validity of the result though. Let's talk about this when we connect on Friday

**Comment by JorgeMartinezG** [comment]
Found the issue. Already sent the fix.
POST request payload needs to be the following:
```
{
	"geotiff_url": "https://odc.ovio.org/?service=WCS&request=GetCoverage&version=1.0.0&coverage=wp_pop_cicunadj&crs=EPSG%3A4326&bbox=92.2%2C9.7%2C101.2%2C28.5&width=1098&height=2304&format=GeoTIFF",
	"zones_url": "https://prism-admin-boundaries.s3.us-east-2.amazonaws.com/mmr_admin_boundaries.json",
	"group_by": "TS",
	"wfs_params": {
		"url": "https://geonode.wfp.org/geoserver/ows",
		"layer_name": "wld_gdacs_tc_events_buffers",
		"time": "2016-05-18",
		"key": "label"
	}
}

```

**Review by dubwalla — COMMENTED** [review]
Looks good from my side. Let's figure out with @ericboucher a merge and deployment plan

File: api-flask/app/zonal_stats.py [hunk 1]
```diff
@@ -16,6 +21,43 @@
 DEFAULT_STATS = ['min', 'max', 'mean', 'median']
 
 
+def get_wfs_response(wfs_params, zones):
+    """Execute wfs request to external service."""
```
Comment (ericboucher) [inline api-flask/app/zonal_stats.py:25]: Let's detail a bit more what this does. Not everyone will be familiar with a "WFS Response"

**Comment by ericboucher** [comment]
@JorgeMartinezG nice work. I just merged a few improvements to the api that created conflicts. Please have a look and take note of what has changed, esp. around typing.

I would recommend adding clearer documentation for this feature, it is not clear what is happening and what info we are expecting from this request.

@avalmas-programize could you have a look at this PR?

Once everything looks good I will test it locally before deploying to the API server.

File: api-flask/app/zonal_stats.py [hunk 2]
```diff
@@ -16,6 +21,43 @@
...
+        'request': 'GetFeature',
+        'typeName': wfs_params.get('layer_name'),
+        'outputFormat': 'application/json',
+        'cql_filter': cql_filter
+    }
+
+    resp = requests.get(wfs_params.get('url'), params)
+    if resp.status_code != 200:
+        logger.error(resp.content)
+        raise ValueError('Invalid response WFS request')
```
[the remark this answers is not in the document]
Comment (JorgeMartinezG) [reply api-flask/app/zonal_stats.py]: Modified message. Included in eef1b51edb81abed1cf780606aec3808cfd50438

File: api-flask/app/zonal_stats.py (hunk 1 above)
Comment (JorgeMartinezG) [reply api-flask/app/zonal_stats.py]: Done. Included in 3661a96c70180094c380a44d99547340ee32811a

**Comment by JorgeMartinezG** [comment]
@ericboucher Applied comments in different commits.
Related to response type checking, there is no defined schema for all layers within a server/database. Even geojson being "strict", properties field might have random key/values.

File: api-flask/app/zonal_stats.py [hunk 3]
```diff
@@ -62,19 +113,80 @@ def _group_zones(zones, group_by):
...
-    geojson_out=False
+    geojson_out=False,
+    wfs_response=None
 ):
     """Calculate stats."""
     if group_by:
         zones = _group_zones(zones, group_by)
 
+    if wfs_response:
+        zones_key = group_by if group_by else 'TS'
```
Comment (dubwalla) [inline api-flask/app/zonal_stats.py:172]: 'TS' is specific to Myanmar. I'm not sure of a use case where the group_by parameter would not be explicitly included in the call to the zonal stats analysis. We should look into making this a mandatory parameter

Comment (JorgeMartinezG) [reply api-flask/app/zonal_stats.py]: I just modified the code in https://github.com/WFP-VAM/prism-frontend/pull/186/commits/7e2378f868750bfe03484e0a364181edcb35cd18 which removes the group_by check and simplifies code drastically. cc @ericboucher

File: api-flask/app/zonal_stats.py [hunk 4]
```diff
@@ -1,26 +1,77 @@
...
+    }
+
+    resp = requests.get(wfs_params.get('url'), params)
+    if resp.status_code != 200:
+        logger.error(resp.content)
+        err_message = 'Received status code from WFS request: {}'.format(resp.status_code)
+
+        raise InternalServerError(err_message)
+
+    # A WFS response should be always a json response.
```
Comment (ericboucher) [inline api-flask/app/zonal_stats.py:66]: -> should always be

Comment (JorgeMartinezG) [reply api-flask/app/zonal_stats.py]: Done. included in e7ba78168bbd18bacd50227d6d28ad7117fdf44a

File: api-flask/app/caching.py [hunk 5]
```diff
@@ -52,6 +52,14 @@ def cache_geojson(geojson, prefix):
     return cache_filepath
 
 
+def get_geojson_file(cached_filepath):
```
Comment (ericboucher) [inline api-flask/app/caching.py:55]: This can open any json, so maybe call it `get_json_file` ?

File: api-flask/app/caching.py [hunk 6]
```diff
@@ -52,6 +52,14 @@ def cache_geojson(geojson, prefix):
...
+    """Return geojson object as python dictionary."""
+    with open(cached_filepath, 'rb') as f:
+        data = json.load(f)
```
Comment (ericboucher) [inline api-flask/app/caching.py:58]: `return json.load(f)` directly?

File: api-flask/app/main.py [hunk 7]
```diff
@@ -92,13 +99,32 @@ def stats():
             url=zones_url
         )
 
+    wfs_params = data.get('wfs_params', None)
+
+    wfs_response = None
```
Comment (ericboucher) [inline api-flask/app/main.py:104]: Let's add a bit more details here in a comment and update the README as well with the new input parameter and its behavior.

Ideally, add some quick tests if it's not too complicated

Then I think we should be good to go :)

Comment (JorgeMartinezG) [reply api-flask/app/main.py:104]: included in https://github.com/WFP-VAM/prism-frontend/pull/186/commits/2f0aa45006bc31f0c60d246aff55f67f4b49f584 and https://github.com/WFP-VAM/prism-frontend/pull/186/commits/286645f27c106a2ca4813612fd135fc77791bfd2

File: api-flask/app/caching.py (hunk 6 above)
Comment (JorgeMartinezG) [reply api-flask/app/caching.py]: Included in https://github.com/WFP-VAM/prism-frontend/pull/186/commits/932e4876ca33afccac1e7460bedb056b04c1ec6c

**Comment by JorgeMartinezG** [comment]
@ericboucher Applied feedback

**Review by ericboucher — APPROVED** [review]

**Pull request merged**